/*
 * Super Resolution
 * Copyright (c) 2025-2026. 187J3X1-114514
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.homo.superresolution.fabric.mixin.compat.iris;

import io.homo.superresolution.common.compat.iris.IrisFramebufferUtils;
import io.homo.superresolution.common.compat.iris.DhDepthReconnectPolicy;
import net.irisshaders.iris.compat.dh.DHCompatInternal;
import net.irisshaders.iris.gl.framebuffer.GlFramebuffer;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.Unique;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(value = DHCompatInternal.class, remap = false)
public class DHCompatInternalMixin {
    @Shadow
    private GlFramebuffer dhWaterFramebuffer;

    @Shadow
    private GlFramebuffer dhTerrainFramebuffer;

    @Inject(method = "reconnectDHTextures", at = @At("RETURN"))
    public void fixDHDepth(int depthTex, CallbackInfo ci) {
        Integer terrainDepth = dhTerrainFramebuffer == null
                ? null : IrisFramebufferUtils.getFramebufferDepthAttachment(dhTerrainFramebuffer.getId());
        Integer waterDepth = dhWaterFramebuffer == null
                ? null : IrisFramebufferUtils.getFramebufferDepthAttachment(dhWaterFramebuffer.getId());
        DhDepthReconnectPolicy.Reconnect reconnect =
                DhDepthReconnectPolicy.evaluate(terrainDepth, waterDepth, depthTex);
        if (reconnect.terrain()) {
            sr$reconnectTerrainDepth(depthTex);
        }
        if (reconnect.water()) {
            sr$reconnectWaterDepth(depthTex);
        }
    }

    @Unique
    private void sr$reconnectTerrainDepth(int depthTex) {
        #if MC_VER < MC_1_21_5
        if (dhTerrainFramebuffer != null) {
            dhTerrainFramebuffer.addDepthAttachment(depthTex);
        }
        #else
        if (dhTerrainFramebuffer != null) {
            dhTerrainFramebuffer.addDepthAttachmentBypass(depthTex);
        }
        #endif
    }

    @Unique
    private void sr$reconnectWaterDepth(int depthTex) {
        #if MC_VER < MC_1_21_5
        if (dhWaterFramebuffer != null) {
            dhWaterFramebuffer.addDepthAttachment(depthTex);
        }
        #else
        if (dhWaterFramebuffer != null) {
            dhWaterFramebuffer.addDepthAttachmentBypass(depthTex);
        }
        #endif
    }
}
